pkg/cmd/admin/router/router.go

```diff
@@ -505,154 +546,183 @@ func RunCmdRouter(f *clientcmd.Factory, cmd *cobra.Command, out io.Writer, cfg *
 
-		objects = app.AddServices(objects, true)
-		// TODO: label all created objects with the same label - router=<name>
-		list := &kapi.List{Items: objects}
+	exposedPorts := make([]kapi.ContainerPort, len(ports))
+	copy(exposedPorts, ports)
+	for i := range exposedPorts {
+		exposedPorts[i].HostPort = 0
+	}
+	containers := []kapi.Container{
+		{
+			Name:            "router",
+			Image:           image,
+			Ports:           exposedPorts,
+			Env:             env.List(),
+			LivenessProbe:   livenessProbe,
+			ReadinessProbe:  readinessProbe,
+			ImagePullPolicy: kapi.PullIfNotPresent,
+			VolumeMounts:    mounts,
+		},
+	}
 
-		if output {
-			list.Items, err = ocmdutil.ConvertItemsForDisplayFromDefaultCommand(cmd, list.Items)
-			if err != nil {
-				return err
-			}
+	if cfg.StatsPort > 0 && cfg.ExposeMetrics {
+		pc := generateMetricsExporterContainer(cfg, env)
+		if pc != nil {
+			containers = append(containers, *pc)
+		}
+	}
 
-			if err := f.PrintObject(cmd, list, out); err != nil {
-				return fmt.Errorf("Unable to print object: %v", err)
+	objects := []runtime.Object{}
+	for _, s := range secrets {
+		objects = append(objects, s)
+	}
+	if createServiceAccount {
+		objects = append(objects,
+			&kapi.ServiceAccount{ObjectMeta: kapi.ObjectMeta{Name: cfg.ServiceAccount}},
+			&authapi.ClusterRoleBinding{
+				ObjectMeta: kapi.ObjectMeta{Name: fmt.Sprintf("router-%s-role", cfg.Name)},
+				Subjects: []kapi.ObjectReference{
+					{
+						Kind:      "ServiceAccount",
+						Name:      cfg.ServiceAccount,
+						Namespace: namespace,
+					},
+				},
+				RoleRef: kapi.ObjectReference{
+					Kind: "ClusterRole",
+					Name: "system:router",
+				},
+			},
+		)
+	}
+	objects = append(objects, &deployapi.DeploymentConfig{
+		ObjectMeta: kapi.ObjectMeta{
+			Name:   name,
+			Labels: label,
+		},
+		Spec: deployapi.DeploymentConfigSpec{
+			Strategy: deployapi.DeploymentStrategy{
+				Type:          deployapi.DeploymentStrategyTypeRolling,
+				RollingParams: &deployapi.RollingDeploymentStrategyParams{UpdatePercent: &updatePercent},
+			},
+			Replicas: cfg.Replicas,
+			Selector: label,
+			Triggers: []deployapi.DeploymentTriggerPolicy{
+				{Type: deployapi.DeploymentTriggerOnConfigChange},
+			},
+			Template: &kapi.PodTemplateSpec{
+				ObjectMeta: kapi.ObjectMeta{Labels: label},
+				Spec: kapi.PodSpec{
+					SecurityContext: &kapi.PodSecurityContext{
+						HostNetwork: cfg.HostNetwork,
+					},
+					ServiceAccountName: cfg.ServiceAccount,
+					NodeSelector:       nodeSelector,
+					Containers:         containers,
+					Volumes:            volumes,
+				},
+			},
+		},
+	})
+
+	objects = app.AddServices(objects, false)
+	// set the service port to the provided hostport value
+	for i := range objects {
+		switch t := objects[i].(type) {
+		case *kapi.Service:
+			for j, servicePort := range t.Spec.Ports {
+				for _, targetPort := range ports {
+					if targetPort.ContainerPort == servicePort.Port && targetPort.HostPort != 0 {
```

Comment:
HostPort is "external", ContainerPort is "internal", so the service port is
going to be the "external" port.  That's roughly equivalent to how host
ports work, which is the closest intent when you aren't using networks.

On Wed, Feb 17, 2016 at 1:03 PM, Michail Kargakis <notifications@github.com>
wrote:

> In pkg/cmd/admin/router/router.go
> <https://github.com/openshift/origin/pull/7251#discussion_r53204452>:
>
> > +					NodeSelector:       nodeSelector,
> > +					Containers:         containers,
> > +					Volumes:            volumes,
> > +				},
> > +			},
> > +		},
> > +	})
> > +
> > +	objects = app.AddServices(objects, false)
> > +	// set the service port to the provided hostport value
> > +	for i := range objects {
> > +		switch t := objects[i].(type) {
> > +		case *kapi.Service:
> > +			for j, servicePort := range t.Spec.Ports {
> > +				for _, targetPort := range ports {
> > +					if targetPort.ContainerPort == servicePort.Port && targetPort.HostPort != 0 {
>
> Shouldn't this compare the container port against the service target port?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/7251/files#r53204452>.
>